npm/qsharp/ux/circuit.tsx

```diff
@@ -12,56 +13,157 @@ import { CircuitProps } from "./data.js";
+      );
+    }
+  }
+
+  function renderCircuit(circuit: qviz.Circuit, container: HTMLDivElement) {
+    qviz.draw(circuit, container);
+
+    // quantum-viz hardcodes the styles in the SVG.
+    // Remove the style elements -- we'll define the styles in our own CSS.
+    const styleElements = container.querySelectorAll("style");
+    styleElements?.forEach((tag) => tag.remove());
+
+    return container.getElementsByClassName("qviz")[0]!;
+  }
+
+  function calculateZoomToFit(container: HTMLDivElement, svg: SVGElement) {
+    const containerWidth = container.clientWidth;
+    // width and height the true dimensions generated by qviz
+    const width = parseInt(svg.getAttribute("width")!);
+    const height = svg.getAttribute("height")!;
+
+    svg.setAttribute("viewBox", `0 0 ${width} ${height}`);
+    const zoom = Math.min(Math.ceil((containerWidth / width) * 100), 100);
+    return zoom;
+  }
+}
+
+function Unrenderable(props: { qubits: number; operations: number }) {
   const errorDiv =
-    props.circuit.qubits.length === 0 ? (
+    props.qubits === 0 ? (
       <div>
         <p>No circuit to display. No qubits have been allocated.</p>
       </div>
-    ) : props.circuit.operations.length > MAX_OPERATIONS ? (
+    ) : props.operations > MAX_OPERATIONS ? (
       <div>
         <p>
-          This circuit has too many gates to display. It has{" "}
-          {props.circuit.operations.length} gates, but the maximum supported is{" "}
-          {MAX_OPERATIONS}.
+          This circuit has too many gates to display. It has {props.operations}{" "}
+          gates, but the maximum supported is {MAX_OPERATIONS}.
         </p>
       </div>
-    ) : props.circuit.qubits.length > MAX_QUBITS ? (
+    ) : props.qubits > MAX_QUBITS ? (
       <div>
         <p>
-          This circuit has too many qubits to display. It has{" "}
-          {props.circuit.qubits.length} qubits, but the maximum supported is{" "}
-          {MAX_QUBITS}.
+          This circuit has too many qubits to display. It has {props.qubits}{" "}
+          qubits, but the maximum supported is {MAX_QUBITS}.
         </p>
       </div>
     ) : undefined;
 
-  useEffect(() => {
-    if (errorDiv !== undefined) {
-      circuitDiv.current!.innerHTML = "";
-      return;
-    }
-
-    qviz.draw(props.circuit, circuitDiv.current!);
-
-    // quantum-viz hardcodes the styles in the SVG.
-    // Remove the style elements -- we'll define the styles in our own CSS.
-    const styleElements = circuitDiv.current?.querySelectorAll("style");
-    styleElements?.forEach((tag) => tag.remove());
-  }, [props.circuit]);
+  return <div class="qs-circuit-error">{errorDiv}</div>;
+}
 
-  return (
-    <div>
-      <div class="qs-circuit-error">{errorDiv}</div>
-      <div class="qs-circuit" ref={circuitDiv}></div>
-    </div>
-  );
+function ZoomControl(props: {
+  zoom?: number;
+  onChange: (zoom: number) => void;
+}) {
+  return props.zoom !== undefined ? (
```

Comment:
Don't need to check for 'undefined' or make zoom prop optional if initialized to a number as per prior comment.

Reply:
OK, got rid of `undefined` , see response to prior comment